### PR TITLE
Rework reconcilation loop. Fix #131.

### LINKED ISSUE #131: Too many replica set are created for registry-ctl component
Run `make sample`

Then check the resource, you'll find more than one registry-ctl replica sets that with 0, 0.
```
replicaset.apps/sample-harbor-registryctl-589b8bb8df   1         1         1       2m16s
replicaset.apps/sample-harbor-registryctl-5bbf7d4889   0         0         0       2m36s
replicaset.apps/sample-harbor-registryctl-644bd8fdbf   0         0         0       2m47s
replicaset.apps/sample-harbor-registryctl-644d9549b6   0         0         0       2m50s
replicaset.apps/sample-harbor-registryctl-687bbd59cd   0         0         0       2m34s
replicaset.apps/sample-harbor-registryctl-6988fbbfbd   0         0         0       2m42s
replicaset.apps/sample-harbor-registryctl-7464d974bb   0         0         0       2m45s
replicaset.apps/sample-harbor-registryctl-74c6d4f8cc   0         0         0       2m35s
replicaset.apps/sample-harbor-registryctl-7569fb84f4   0         0         0       2m52s
replicaset.apps/sample-harbor-registryctl-79cc6ccbf    0         0         0       2m46s
replicaset.apps/sample-harbor-registryctl-c58b8b4d8    0         0         0       2m56s
```

Deployment is not ready:

```
NAME                                        READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/sample-harbor-registryctl   0/1     1            0           2d3h
```

### PR DESCRIPTION
**Pull request by sguyennet**

### CODE REVIEW

**Comment by steven-zou** [comment]
@sguyennet 

Thanks for this fix. Would you please check and fix the failed CI checks? e.g: golint, generated files etc..

**Comment by bitsf** [comment]
hi @sguyennet Can you also point out which lines in the fix is relative to reconcilation loop problems, that can help us review

**Comment by holyhope** [comment]
There is at least 2 issues fixed in this PR:

1. Check that the resources is ready before applying the next one. (the check was broken because not called every time, so resource may be applied even if a dependency is not ready).
   Solved by reworking `pkg/controller/resource.ProcessFunc`.
2. The resource is changed on every apply because controller apply the template omitting fields in order to be filled by apiserver with default values. using this method, the object is modifiedon each apply because the `.metadata.managedFields` is updated.
   Solved by using [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) instead of [`CreateOrUpdate`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L197).
   Using server-side apply leads to many code removal since the *merge* of resources is done by kubernetes.
3. Removed the check before reconciling resources to skip the loop if no changes detected. See (`pkg/controller/status.go`).
   There was a check to ignore some reconciliation events to *optimize* the controller. By design, this check cannot detect changes properly.

**Comment by steven-zou** [comment]
@sguyennet @holyhope 

CI is still failing. Would you please check and fix the failures ASAP? Thanks a lot!

**Comment by steven-zou** [comment]
@sguyennet

Any updates about this PR (When can we fix the CI failures)? We're heading to the FC deadline.

File: controllers/goharbor/internal/test/certificate.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
+/*
```
Comment (holyhope) [inline controllers/goharbor/internal/test/certificate.go:1]: IMO you should use fucntions from `controllers/goharbor/test_utils.go`, there is bugs here :-)

**Review by holyhope — CHANGES_REQUESTED** [review]

**Review by holyhope — APPROVED** [review]

**Review by steven-zou — APPROVED** [review]
LGTM

**Pull request merged**